### PR TITLE
libr/bin: completed the heuristic to find main for mips binaries

### PR DESCRIPTION
**Pull request by dkreuter**
tested with the binary in #2074

### CODE REVIEW

File: libr/bin/format/elf/elf.c [hunk 1]
```diff
@@ -488,12 +488,40 @@ ut64 Elf_(r_bin_elf_get_main_offset)(struct Elf_(r_bin_elf_obj_t) *bin) {
...
+		{
+			return 0;
+		}
+
+		ut64 gp = got_offset + 0x7ff0;
+
+		unsigned i;
+		for (i=0; i < sizeof(buf)/sizeof(buf[0]); i+=4) {
+			#define BUF_U32(i) ((ut32)(buf[i+0]+(buf[i+1]<<8)+(buf[i+2]<<16)+(buf[i+3]<<24)))
+			ut32 instr = BUF_U32(i);
```
Comment (jvoisin) [inline libr/bin/format/elf/elf.c]: +`const`

File: libr/bin/format/elf/elf.c [hunk 2]
```diff
@@ -488,12 +488,40 @@ ut64 Elf_(r_bin_elf_get_main_offset)(struct Elf_(r_bin_elf_obj_t) *bin) {
...
+
+		ut64 got_offset;
+
+		if ((got_offset = Elf_(r_bin_elf_get_section_offset) (bin, ".got")) == -1 &&
+		    (got_offset = Elf_(r_bin_elf_get_section_offset) (bin, ".got.plt")) == -1)
+		{
+			return 0;
+		}
+
+		ut64 gp = got_offset + 0x7ff0;
```
Comment (jvoisin) [inline libr/bin/format/elf/elf.c]: Why `0x7ff0`? Please add a comment :)

File: libr/bin/format/elf/elf.c [hunk 3]
```diff
@@ -488,12 +488,40 @@ ut64 Elf_(r_bin_elf_get_main_offset)(struct Elf_(r_bin_elf_obj_t) *bin) {
...
+
+		unsigned i;
```
Comment (radare) [inline libr/bin/format/elf/elf.c]: Dont use undefined types. Plain int should be fine. But you can use ut32 if really 2GB Limit matters

File: libr/bin/format/elf/elf.c [hunk 4]
```diff
@@ -488,12 +488,40 @@ ut64 Elf_(r_bin_elf_get_main_offset)(struct Elf_(r_bin_elf_obj_t) *bin) {
...
+		/*
+		    assuming the startup code looks like
+		        got = gp-0x7ff0
+		        got[index__libc_start_main] ( got[index_main] );
+
+		    looking for the instruction generating the first argument to find main
+		        lw a0, offset(gp)
+		*/
+
+		ut64 got_offset;
```
Comment (radare) [inline libr/bin/format/elf/elf.c]: Do not define vars in the middle of nowhwre. Use stackframes/brackets

File: libr/bin/format/elf/elf.c [hunk 5]
```diff
@@ -488,12 +488,40 @@ ut64 Elf_(r_bin_elf_get_main_offset)(struct Elf_(r_bin_elf_obj_t) *bin) {
...
+
+		ut64 gp = got_offset + 0x7ff0;
+
+		unsigned i;
+		for (i=0; i < sizeof(buf)/sizeof(buf[0]); i+=4) {
+			#define BUF_U32(i) ((ut32)(buf[i+0]+(buf[i+1]<<8)+(buf[i+2]<<16)+(buf[i+3]<<24)))
+			ut32 instr = BUF_U32(i);
+			if ((instr & 0xffff0000) == 0x8f840000) { // lw a0, offset(gp)
+				short delta = instr;
+				r_buf_read_at (bin->b, /* got_entry_offset = */ gp + delta, buf, 4);
```
Comment (radare) [inline libr/bin/format/elf/elf.c]: Short requires some implicit casting iirc

**Comment by radare** [comment]
Thanks!

**Pull request closed without merging**